File: ambeth/Ambeth.Util/ambeth/collections/CleanupInvalidKeysSet.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace De.Osthus.Ambeth.Collections
{
    public class CleanupInvalidKeysSet<K> : CHashSet<K>
    {
        protected readonly IInvalidKeyChecker<K> invalidKeyChecker;

	    protected int modCountDownCount;

        public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker)
            : base()
	    {
		    this.invalidKeyChecker = invalidKeyChecker;
            ResetCounter();
	    }

        public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker, ICollection<K> sourceCollection)
            : base(sourceCollection)
	    {
		    this.invalidKeyChecker = invalidKeyChecker;
            ResetCounter();
	    }

        public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker, float loadFactor)
            : base(loadFactor)
	    {
		    this.invalidKeyChecker = invalidKeyChecker;
            ResetCounter();
	    }

        public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker, IList sourceCollection)
            : base(sourceCollection)
        {
            this.invalidKeyChecker = invalidKeyChecker;
            ResetCounter();
        }

        public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker, IList<K> sourceCollection)
            : base(sourceCollection)
        {
            this.invalidKeyChecker = invalidKeyChecker;
            ResetCounter();
        }

        public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker, K[] sourceArray)
            : base(sourceArray)
        {
            this.invalidKeyChecker = invalidKeyChecker;
            ResetCounter();
        }

        public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker, int initialCapacity)
            : base(initialCapacity)
        {
            this.invalidKeyChecker = invalidKeyChecker;
            ResetCounter();
        }

        public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker, int initialCapacity, float loadFactor)
            : base(initialCapacity, loadFactor)
        {
            this.invalidKeyChecker = invalidKeyChecker;
            ResetCounter();
        }

        protected override void EntryAdded(SetEntry<K> entry)
        {
 	         base.EntryAdded(entry);
		     SetChanged();
	    }

        protected override void EntryRemoved(SetEntry<K> entry)
        {
 	         base.EntryRemoved(entry);
             SetChanged();
	    }

        protected void ResetCounter()
        {
            modCountDownCount = threshold / 2 + 1;
        }

	    protected void SetChanged()
	    {
		    if (--modCountDownCount != 0)
		    {
			    return;
		    }
		    // size() can never be 0 after super.entryAdded(). So this statement is only true if we want the "one-time" cleanup operation
		    // between two transfer() calls
            modCountDownCount = 0; // set the counter to zero to suppress triggering of "setChanged()" from entryRemoved()
            Transfer(table);
            ResetCounter();
	    }

        protected override void Resize(int newCapacity)
        {
            modCountDownCount = 0; // set the counter to zero to suppress triggering of "setChanged()" from entryRemoved()
            base.Resize(newCapacity);
            ResetCounter();
        }

        protected override bool IsEntryValid(De.Osthus.Ambeth.Collections.SetEntry<K> entry)
        {
            return invalidKeyChecker.IsKeyValid(entry.Key);
	    }
    }
}

File: jambeth/jambeth-util/src/main/java/com/koch/ambeth/util/collections/CleanupInvalidKeysSet.java
package com.koch.ambeth.util.collections;

/*-
 * #%L
 * jambeth-util
 * %%
 * Copyright (C) 2017 Koch Softwaredevelopment
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 * #L%
 */

import java.util.Collection;
import java.util.Set;

public class CleanupInvalidKeysSet<K> extends HashSet<K> {
	protected final IInvalidKeyChecker<K> invalidKeyChecker;

	protected int modCountDownCount;

	public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker) {
		super();
		this.invalidKeyChecker = invalidKeyChecker;
		resetCounter();
	}

	public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker,
			Collection<? extends K> sourceCollection) {
		super(sourceCollection);
		this.invalidKeyChecker = invalidKeyChecker;
		resetCounter();
	}

	public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker, float loadFactor) {
		super(loadFactor);
		this.invalidKeyChecker = invalidKeyChecker;
		resetCounter();
	}

	@SuppressWarnings("rawtypes")
	public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker, int initialCapacity,
			float loadFactor, Class<? extends ISetEntry> entryClass) {
		super(initialCapacity, loadFactor, entryClass);
		this.invalidKeyChecker = invalidKeyChecker;
		resetCounter();
	}

	public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker, int initialCapacity,
			float loadFactor) {
		super(initialCapacity, loadFactor);
		this.invalidKeyChecker = invalidKeyChecker;
		resetCounter();
	}

	public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker, int initialCapacity) {
		super(initialCapacity);
		this.invalidKeyChecker = invalidKeyChecker;
		resetCounter();
	}

	public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker, K[] sourceArray) {
		super(sourceArray);
		this.invalidKeyChecker = invalidKeyChecker;
		resetCounter();
	}

	public CleanupInvalidKeysSet(IInvalidKeyChecker<K> invalidKeyChecker, Set<? extends K> map) {
		super(map);
		this.invalidKeyChecker = invalidKeyChecker;
		resetCounter();
	}

	@Override
	protected void entryAdded(ISetEntry<K> entry) {
		super.entryAdded(entry);
		setChanged();
	}

	@Override
	protected void entryRemoved(ISetEntry<K> entry) {
		super.entryRemoved(entry);
		setChanged();
	}

	protected void resetCounter() {
		modCountDownCount = threshold / 2 + 1;
	}

	protected void setChanged() {
		if (--modCountDownCount != 0) {
			return;
		}
		// size() can never be 0 after super.entryAdded(). So this statement is only true if we want the
		// "one-time" cleanup operation
		// between two transfer() calls
		modCountDownCount = 0; // set the counter to zero to suppress triggering of "setChanged()" from
														// entryRemoved()
		transfer(table);
		resetCounter();
	}

	@Override
	protected void resize(int newCapacity) {
		modCountDownCount = 0; // set the counter to zero to suppress triggering of "setChanged()" from
														// entryRemoved()
		super.resize(newCapacity);
		resetCounter();
	}

	@Override
	protected boolean isEntryValid(ISetEntry<K> entry) {
		return invalidKeyChecker.isKeyValid(entry.getKey());
	}
}
